File: problems/algoexpert/river_sizes.js
const matrix = [
  [1, 0, 0, 1, 0],
  [1, 0, 1, 0, 0],
  [0, 0, 1, 0, 1],
  [1, 0, 1, 0, 1],
  [1, 0, 1, 1, 0]
];

/**
 *  Iterate over the matrix
 * $$ treat every point like the entry node for a graph traversal. Each node is an array of coords $$
 * for each node, if its not visited AND if its start of a river, get its neighbours using BFS or DFS
 *  incrementing the current river size if the neighbour is unvisited AND is == 1 (river), then get its neighbours and add to Q
 */

// BFS solution

function riverSizes(matrix) {
  let res = [];
  let visited = matrix.map(row => row.map(cell => false));
  // let visited =  new Array(matrix.length).fill(new Array (matrix[0].length).fill(false))

  // iterate over the matrix
  for (let r = 0; r < matrix.length; r++) {
    for (let c = 0; c < matrix[0].length; c++) {
      if (!visited[r][c]) {
        visit(r, c, matrix, visited, res);
      }
    }
  }

  return res;
}

console.log(riverSizes(matrix));

// BFS implementation
function visit(rowIndex, colIndex, matrix, visited, res) {
  // track current river size if current node is start of river
  let currentRiverSize = 0;

  // set up Q for BFS
  let Q = [[rowIndex, colIndex]]; // each element is an array of coordinates

  // BFS
  while (Q.length > 0) {
    const currentNode = Q.pop();

    let [r, c] = currentNode;
    if (!visited[r][c]) {
      // mark visited
      visited[r][c] = true;
      // check if river
      if (matrix[r][c] === 1) {
        currentRiverSize++;
        // add neighbours to Q
        let neighbours = getNeighbours(r, c, matrix, visited);
        neighbours.forEach(neighbour => {
          Q.unshift(neighbour);
        });
      }
    } else {
    }
  }
  // after Q is empty,you have the river size for that node
  if (currentRiverSize > 0) res.push(currentRiverSize);
}

function getNeighbours(rowInd, colInd, matrix, visited) {
  let neighbours = [];
  // get prev row and next row for [r,c] => up and down
  let prevRow = rowInd - 1;
  let nextRow = rowInd + 1;
  if (prevRow >= 0) {
    // if (prevRow >= 0 && !visited[prevRow][colInd]) {
    neighbours.push([prevRow, colInd]);
  }

  if (nextRow < matrix.length)
    // if (nextRow < matrix.length && !visited[nextRow][colInd])
    neighbours.push([nextRow, colInd]);

  // get prev col and next col => left and right
  let prevCol = colInd - 1;
  let nextCol = colInd + 1;
  if (prevCol >= 0)
    // if (prevCol >= 0 && !visited[rowInd][prevCol])
    neighbours.push([rowInd, prevCol]); // push coordinates
  if (nextCol < matrix[0].length)
    // if (nextCol < matrix[0].length && !visited[rowInd][nextCol])
    neighbours.push([rowInd, nextCol]); // push coordinates

  return neighbours;
}

function $getNeighbours(r, c, grid) {
  let neighbours = [];
  // first and last rows
  if (r > 0) {
    let top = [r - 1, c];
    neighbours.push(top);
  }
  if (r < grid.length - 1) {
    let bottom = [r + 1, c];
    neighbours.push(bottom);
  }
  // left and right cols
  if (c > 0) {
    let left = [r, c - 1];
    neighbours.push(left);
  }
  if (c < grid[0].length - 1) {
    let right = [r, c + 1];
    neighbours.push(right);
  }

  return neighbours;
}

File: problems/algoexpert/python/river_sizes.py
matrix = [
    [1, 0, 0, 1, 0],
    [1, 0, 1, 0, 0],
    [0, 0, 1, 0, 1],
    [1, 0, 1, 0, 1],
    [1, 0, 1, 1, 0]
] # ans = [2, 1, 5, 2, 2]


def riverSizes(matrix):
    """
      1) create visited matrix => a copy of the matrix, with each value = False
      2) iterate over main matrix, and for each value in it, visit(r,c) it
      3) visit() does the following:
          iA) add the r,c pair into a Q (BFS), then immediate pop
          iB) initialise currentRiverSize = 0
          ii) While Q is !empty,  if the matrix at[r,c]  is unvisited:
              (iii) if it is also === 1:
                  (iv) increment the currentRiverSize by 1 
                  (v) get all its neighbours (top, bottom, left, right)
                  (vi) add neighbours to the Q
          (viii) now that Q is empty, if currentRiverSize  > 0 add it to the results list
      4) getNeighbours() does the following:
        (i) gets top and bottom neighbours - prev row and next row
        (ii) gets left and right neighbours - prev col and next col
        (iii) inserts them as pairs of points [r,c] into the returning array which gets iterated over in 3(vi)
    """

    res = []
    visited = [[False for cell in row] for row in matrix]

    # iterate over the entire matrix, and for each
    # coord/node run a BFS
    for r in range(len(matrix)):
        for c in range(len(matrix[0])):
            if not visited[r][c]:
                visit(r, c, matrix, visited, res)

    return res


def visit(row, col, matrix, visited, res):
    currentRiverLen = 0
    Q = [[row, col]]

    while len(Q) > 0:
        currentNode = Q.pop()
        r = currentNode[0]
        c = currentNode[1]

        if not visited[r][c]:
            visited[r][c] = True
            if matrix[r][c] == 1:
                currentRiverLen += 1
                neighbours = getNeighbours(r, c, matrix, visited)
                for neighbour in neighbours:
                    Q.insert(0, neighbour)

    # When Q empty
    if currentRiverLen > 0:
        res.append(currentRiverLen)


def getNeighbours(row, col, matrix, visited):
    neighbours = []
    # check prev row and next row = up and down from current point
    if row-1 >= 0:
        neighbours.append([row-1, col])
    if row + 1 < len(matrix):
        neighbours.append([row+1, col])

    # columns => left and right of the current point
    if col-1 >= 0:
        neighbours.append([row, col-1])
    if col+1 < len(matrix[0]):
        neighbours.append([row, col+1])

    return neighbours


ans = riverSizes(matrix)

print(ans)
